### PR TITLE
[BEAM-5959] DO NOT MERGE: KMS support for FileBasedSink, GCS

### PR DESCRIPTION
**Pull request by udim**
Adds kmsKey setting to file creation (create, copy) for FileBasedSink IOs.
Supported transforms: FileIO, TextIO, AvroIO, TFRecordIO

Not yet ready:
- Requires google_cloud_bigdataoss_version = "1.9.12" (or later)
- Still missing some tests.

Note to reviewers: I've tried to maintain backwards compatibility in public APIs. Please verify.

------------------------

Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

It will help us expedite review of your Pull Request if you tag someone (e.g. `@username`) to look at it.

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | --- | --- | --- | ---
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_Verify/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/) </br> [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | --- | --- | --- | ---

### CODE REVIEW

**Comment by udim** [comment]
Ready for initial reviews. R: @chamikaramj , @iemejia

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/FileSystem.java [hunk 1]
```diff
@@ -102,8 +107,52 @@ protected abstract WritableByteChannel create(ResourceIdT resourceId, CreateOpti
...
+   * @param destResourceIds the references of the destination resources
+   * @param moveOptions additional settings passed to the underlying implementation
+   * @throws FileNotFoundException if the source resources are missing. When copy throws, each
+   *     resource might or might not be copied. In such scenarios, callers can use {@code match()}
+   *     to determine the state of the resources.
+   */
+  protected void copy(
+      List<ResourceIdT> srcResourceIds, List<ResourceIdT> destResourceIds, MoveOptions moveOptions)
+      throws IOException {
+    LOG.warn("Using deprecated copy");
```
Comment (iemejia) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/FileSystem.java]: This is not the deprecated one, no? I mean should not log here.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/FileSystem.java [hunk 2]
```diff
@@ -102,8 +107,52 @@ protected abstract WritableByteChannel create(ResourceIdT resourceId, CreateOpti
...
+   * resources will be created recursively.
+   *
+   * @param srcResourceIds the references of the source resources
+   * @param destResourceIds the references of the destination resources
+   * @param moveOptions additional settings passed to the underlying implementation
+   * @throws FileNotFoundException if the source resources are missing. When copy throws, each
+   *     resource might or might not be copied. In such scenarios, callers can use {@code match()}
+   *     to determine the state of the resources.
+   */
+  protected void copy(
```
Comment (iemejia) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/FileSystem.java]: I kinda see where you are going with having an Options object for file operations, But we should probably think more in detail what are this options for, e.g. having `MoveOptions` all around make sense ?

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/fs/MoveOptions.java [hunk 3]
```diff
@@ -17,16 +17,61 @@
...
+  public abstract boolean getIgnoreMissingFiles();
+
+  /**
+   * Specifies the Key Management System key name to use to encrypt new files with.
+   *
+   * <p>Only relevant to operations that create new files and filesystems that support KMS features.
+   */
+  @Experimental(Kind.FILESYSTEM)
+  @Nullable
+  public abstract String getDestKmsKey();
```
Comment (iemejia) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/fs/MoveOptions.java]: This should not be done here but better in the FileSystem options. Notice also that you may need to set up a different SourceKmsKey so this should be considered.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/FileSystem.java (hunk 1 above)
Comment (udim) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/io/FileSystem.java]: The idea is to add a warning to the implementer of a FileSystem that's not in the Beam repo to implement this method. I could remove it if it's too spammy.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/FileSystem.java (hunk 2 above)
Comment (udim) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/io/FileSystem.java]: Here I was copying CreateOptions, which is passed to `FileSystem.create()` implementations.
Were you thinking of having seperate FooOptions classes for each operation? Or a single class for all?

File: buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy [hunk 4]
```diff
@@ -324,7 +324,8 @@ class BeamModulePlugin implements Plugin<Project> {
     // Maven artifacts.
     def generated_grpc_beta_version = "0.19.0"
     def generated_grpc_ga_version = "1.18.0"
-    def google_cloud_bigdataoss_version = "1.9.0"
+    // TODO: do not merge this change
+    def google_cloud_bigdataoss_version = "1.9.12-SNAPSHOT"
```
[the remark this answers is not in the document]
Comment (udim) [reply buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy]: Yeah, I'm using unreleased locally compiled code from HEAD here.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/AvroIO.java [hunk 5]
```diff
@@ -431,7 +431,8 @@
         .setNumShards(0)
         .setCodec(TypedWrite.DEFAULT_SERIALIZABLE_CODEC)
         .setMetadata(ImmutableMap.of())
-        .setWindowedWrites(false);
+        .setWindowedWrites(false)
+        .setKmsKey(null);
```
[the remark this answers is not in the document]
Comment (udim) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/io/AvroIO.java]: From what I can tell S3Options are set up from command line flags. Is there a way to use different awsKmsKeyIds for 2 or more sinks in the same pipeline? That was my goal.

**Comment by udim** [comment]
@iemejia I've considered this further and I believe that we don't really have a use case (yet) for multiple KMS keys per pipeline. In the interest of simplicity I'll revert this PR and create a new one that only touches GCP/GCS.

**Comment by udim** [comment]
The idea is to register a new filesystem with different options. That could work, yes. Thanks for the pointer!

**Pull request closed without merging**